### PR TITLE
[Feature] Support message trace feature

### PR DESCRIPTION
**Pull request by hugoasdf**
## What is the purpose of the change

[Feature] Support message trace feature

## Brief changelog

add message trace feature for cpp sdk,add hook,dispatcher

## Verifying this change

can send message trace message.
TBD:web console show.

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when a cross-module dependency exists.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: src/consumer/MQConsumer.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "MQConsumer.h"
+
+#include "Logging.h"
+#include "ConsumeMessageHook.h"
+#include "ClientRPCHook.h"
+#include "trace/hook/ConsumeMessageTraceHookImpl.h"
+#include "AsyncTraceDispatcher.h"
+namespace rocketmq {
+
+      MQConsumer::MQConsumer(PullRequest* request) {
```
Comment (ifplusor) [inline src/consumer/MQConsumer.cpp]: where the param request is used?

File: include/MQConsumer.h [hunk 2]
```diff
@@ -59,8 +82,25 @@ class ROCKETMQCLIENT_API MQConsumer : public MQClient {
 
  protected:
   MessageModel m_messageModel;
+
+
+  private:
+  std::shared_ptr<TraceDispatcher> traceDispatcher;
```
Comment (ifplusor) [inline include/MQConsumer.h]: rename to 'm_traceDispatcher'

File: include/MQConsumer.h [hunk 3]
```diff
@@ -59,8 +82,25 @@ class ROCKETMQCLIENT_API MQConsumer : public MQClient {
...
   MessageModel m_messageModel;
+
+
+  private:
+  std::shared_ptr<TraceDispatcher> traceDispatcher;
+  protected:
+   //DefaultMQProducerImpl defaultMQProducerImpl;
+
+  
+  std::vector<std::shared_ptr<ConsumeMessageHook> > consumeMessageHookList;
```
Comment (ifplusor) [inline include/MQConsumer.h]: same as above.

File: src/consumer/MQConsumer.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+				catch (...) {
+					LOG_ERROR("system mqtrace hook init failed ,maybe can't send msg trace data");
+				}
+			}//if
+
+        }
+
+
+   MQConsumer::~MQConsumer() {
+	if (traceDispatcher.use_count()>0) {
```
Comment (ifplusor) [inline src/consumer/MQConsumer.cpp]: don't use 'use_count() > 0' as condition, but 'traceDispatcher != nullptr'.

File: src/producer/DefaultMQProducer.cpp [hunk 5]
```diff
@@ -32,11 +32,17 @@
 #include "BatchMessage.h"
 #include <typeinfo>
 
+
+#include "AsyncTraceDispatcher.h"
+#include <boost/shared_ptr.hpp>
```
Comment (ifplusor) [inline src/producer/DefaultMQProducer.cpp]: where use boost::shared_ptr or boost::weak_ptr ?

File: include/AsyncTraceDispatcher.h [hunk 6]
```diff
@@ -0,0 +1,165 @@
+#ifndef __AsyncTraceDispatcher_H__
```
Comment (ifplusor) [inline include/AsyncTraceDispatcher.h]: add Apache License declaration.

File: include/AsyncTraceDispatcher.h [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+struct AsyncRunnable_run_context {
+  //bool stopped;
+  int batchSize;
+  //AsyncTraceDispatcher* atd;
+  std::shared_ptr<AsyncTraceDispatcher> atd;
+  //DefaultMQProducer* traceProducer;
+  std::string TraceTopicName;
+  AsyncRunnable_run_context(bool stoppedv, int batchSizev,
+	  //AsyncTraceDispatcher* atdv,
```
Comment (ifplusor) [inline include/AsyncTraceDispatcher.h]: format all changed files with clang-format.

File: src/trace/hook/SendMessageTraceHookImpl.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+#include "SendMessageHook.h"
+namespace rocketmq {
+class SendMessageTraceHookImpl : public SendMessageHook {
+ private:
+  std::shared_ptr<TraceDispatcher> localDispatcher;
+
+  public:
+  SendMessageTraceHookImpl(std::shared_ptr<TraceDispatcher>& localDispatcher);
+  virtual std::string hookName();
+  //virtual void sendMessageBefore(SendMessageContext* context);
```
Comment (ifplusor) [inline src/trace/hook/SendMessageTraceHookImpl.h]: remove unnecessary functions.

File: src/trace/hook/ConsumeMessageTraceHookImpl.cpp [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+			TraceContext subAfterContext;
+			subAfterContext.setTraceType(TraceType::SubAfter);//
+			subAfterContext.setRegionId(subBeforeContext->getRegionId());//
+			subAfterContext.setGroupName("NamespaceUtil.withoutNamespace(subBeforeContext.getGroupName())");//
+			subAfterContext.setRequestId(subBeforeContext->getRequestId());//
+      subAfterContext.setSuccess(context ->isSuccess());                                              //
+
+			// Caculate the cost time for processing messages
+			int costTime = 0;//(int)((System.currentTimeMillis() - subBeforeContext.getTimeStamp()) / context.getMsgList().size());
+			subAfterContext.setCostTime(costTime);//
```
Comment (ifplusor) [inline src/trace/hook/ConsumeMessageTraceHookImpl.cpp]: calculate real costTime.

File: src/trace/hook/ConsumeMessageTraceHookImpl.cpp [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+package org.apache.rocketmq.client.trace.hook;
```
Comment (ifplusor) [inline src/trace/hook/ConsumeMessageTraceHookImpl.cpp]: remove unnecessary code.

Comment (hugoasdf) [reply src/trace/hook/ConsumeMessageTraceHookImpl.cpp]: review recommends are read,checks will become,thanks.

> 
> 
> remove unnecessary code.

File: include/AsyncTraceDispatcher.h [hunk 11]
```diff
@@ -0,0 +1,118 @@
+#ifndef __AsyncTraceDispatcher_H__
```
Comment (jonnxu) [inline include/AsyncTraceDispatcher.h]: please use big alphabet for macro definition,  the same for following files
__AsyncTraceDispatcher_H__ change to __ASYNCTRACEDISPATCHER_H__

File: include/AsyncTraceDispatcher.h [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+#include "TraceHelper.h"
+
+#include <memory>
+#include "ClientRPCHook.h"
+#include "TraceDispatcher.h"
+namespace rocketmq {
+
+class AsyncTraceDispatcher : public TraceDispatcher, public enable_shared_from_this<AsyncTraceDispatcher> {
+ private:
+  // static InternalLogger log;  //= ClientLogger.getLog();
```
Comment (jonnxu) [inline include/AsyncTraceDispatcher.h]: delete the unnecessary code for new class

File: include/AsyncTraceDispatcher.h [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+  std::list<TraceContext> traceContextQueue;
+
+  // ArrayBlockingQueue<Runnable> appenderQueue;
+  std::list<TraceContext> appenderQueue;
+
+  std::mutex m_appenderQueuenotEmpty_mutex;
+  std::condition_variable m_appenderQueuenotEmpty;
+
+  std::thread* m_shutDownHook;
+  bool m_stopped = false;
```
Comment (jonnxu) [inline include/AsyncTraceDispatcher.h]: not assignment class variant here

File: include/AsyncTraceDispatcher.h [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+  int m_queueSize;
+  int m_batchSize;
+  int m_maxMsgSize;
+  // trace message Producer
+  // ThreadPoolExecutor traceExecutor;
+  std::atomic<long> m_discardCount;
+  // std::thread* worker;
+  std::shared_ptr<std::thread> m_worker;
+
+ public:
```
Comment (jonnxu) [inline include/AsyncTraceDispatcher.h]: why here is public class members ?

File: include/ConsumeMessageHook.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+#define __ConsumeMessageHook_H__
+
+#include <string>
+#include "ConsumeMessageContext.h"
+
+namespace rocketmq {
+
+class ConsumeMessageHook {
+public:
+	virtual std::string hookName() { return ""; }
```
Comment (jonnxu) [inline include/ConsumeMessageHook.h]: here format, please check if there is format problem for others

File: include/DefaultMQProducer.h [hunk 16]
```diff
@@ -28,7 +28,7 @@ namespace rocketmq {
 //<!***************************************************************************
 class ROCKETMQCLIENT_API DefaultMQProducer : public MQProducer {
  public:
-  DefaultMQProducer(const std::string& groupname);
+  DefaultMQProducer(const std::string& groupname, bool Withouttrace = false, void* rpcHook = nullptr);
```
Comment (jonnxu) [inline include/DefaultMQProducer.h]: notice the name format for funtion variant, such as Withouttrace, please check others

File: include/TraceHelper.h [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+  SubAfter,
+};
+
+
+class TraceBean {
+ public:
+  static std::string LOCAL_ADDRESS;  //= "/*UtilAll.ipToIPv4Str(UtilAll.getIP())*/";
+  
+ private:
+  std::string m_topic = "";
```
Comment (jonnxu) [inline include/TraceHelper.h]: no need assignment here

File: src/message/MQMessage.cpp [hunk 18]
```diff
@@ -46,6 +46,9 @@ const string MQMessage::PROPERTY_CONSUME_START_TIMESTAMP = "CONSUME_START_TIME";
 const string MQMessage::PROPERTY_TRANSACTION_PREPARED_QUEUE_OFFSET = "TRAN_PREPARED_QUEUE_OFFSET";
 const string MQMessage::PROPERTY_TRANSACTION_CHECK_TIMES = "TRANSACTION_CHECK_TIMES";
 const string MQMessage::PROPERTY_CHECK_IMMUNITY_TIME_IN_SECONDS = "CHECK_IMMUNITY_TIME_IN_SECONDS";
+const string MQMessage::CONSUME_CONTEXT_TYPE = "ConsumeContextType";
```
Comment (jonnxu) [inline src/message/MQMessage.cpp]: keep the same style for parameter name

Comment (hugoasdf) [reply src/message/MQMessage.cpp]: > 
> 
> keep the same style for parameter name

updated,-:)

File: include/AsyncTraceDispatcher.h [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+#include "DefaultMQProducer.h"
+
+#include "TraceHelper.h"
+
+#include <memory>
+#include "ClientRPCHook.h"
+#include "TraceDispatcher.h"
+namespace rocketmq {
+
+class AsyncTraceDispatcher : public TraceDispatcher, public enable_shared_from_this<AsyncTraceDispatcher> {
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]: `public enable_shared_from_this<AsyncTraceDispatcher>`  seems can be `private` here right?

File: include/AsyncTraceDispatcher.h [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+ private:
+  int m_queueSize;
+  int m_batchSize;
+  int m_maxMsgSize;
+  std::atomic<long> m_discardCount;
+  std::shared_ptr<std::thread> m_worker;
+
+  // public:
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]: remove this line?

File: include/AsyncTraceDispatcher.h [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+  std::thread* m_shutDownHook;
+  bool m_stopped;
+  std::string m_dispatcherId;
+  std::string m_traceTopicName;
+  std::atomic<bool> m_isStarted;
+  AccessChannel m_accessChannel;
+  std::atomic<bool> m_delydelflag;
+
+ public:
+  bool get_stopped() { return m_stopped; }
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]:
```suggestion
  bool get_stopped() const { return m_stopped; }
```

File: include/AsyncTraceDispatcher.h [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+
+  virtual bool append(TraceContext* ctx);
+  virtual void flush();
+  virtual void shutdown();
+  virtual void registerShutDownHook();
+  virtual void removeShutdownHook();
+  virtual void start(std::string nameSrvAddr, AccessChannel accessChannel = AccessChannel::LOCAL);
+
+  virtual void setdelydelflag(bool v) { m_delydelflag = v; }
+  bool getdelydelflag() { return m_delydelflag; }
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]:
```suggestion
  bool getdelydelflag() const { return m_delydelflag; }
```

File: include/AsyncTraceDispatcher.h [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+
+  AccessChannel getAccessChannel() { return m_accessChannel; }
+
+  void setAccessChannel(AccessChannel accessChannelv) { m_accessChannel = accessChannelv; }
+
+  std::string& getTraceTopicName() { return m_traceTopicName; }
+
+  void setTraceTopicName(std::string traceTopicNamev) { m_traceTopicName = traceTopicNamev; }
+  bool getisStarted() { return m_isStarted.load(); };
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]:
```suggestion
  bool getisStarted() const { return m_isStarted.load(); };
```

File: include/AsyncTraceDispatcher.h [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+  virtual void start(std::string nameSrvAddr, AccessChannel accessChannel = AccessChannel::LOCAL);
+
+  virtual void setdelydelflag(bool v) { m_delydelflag = v; }
+  bool getdelydelflag() { return m_delydelflag; }
+
+  AccessChannel getAccessChannel() { return m_accessChannel; }
+
+  void setAccessChannel(AccessChannel accessChannelv) { m_accessChannel = accessChannelv; }
+
+  std::string& getTraceTopicName() { return m_traceTopicName; }
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]:
```suggestion
  std::string& getTraceTopicName() const { return m_traceTopicName; }
```

File: include/AsyncTraceDispatcher.h [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+
+  void setTraceTopicName(std::string traceTopicNamev) { m_traceTopicName = traceTopicNamev; }
+  bool getisStarted() { return m_isStarted.load(); };
+
+ std::shared_ptr<DefaultMQProducer> getTraceProducer() { return m_traceProducer; }
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]:
```suggestion
 std::shared_ptr<DefaultMQProducer> getTraceProducer() const { return m_traceProducer; }
```

File: include/AsyncTraceDispatcher.h [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+struct AsyncRunnable_run_context {
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]: The name of this class seems weird.

File: include/AsyncTraceDispatcher.h [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+  void sendTraceDataByMQ(std::vector<std::string> keySet,
+                         std::string data,
+                         std::string dataTopic,
+                         std::string regionId);
+  std::set<std::string> tryGetMessageQueueBrokerSet(std::shared_ptr<DefaultMQProducer> producer, std::string topic);
+};
+
+}  // namespace rocketmq
+
+#endif
```
Comment (jovany-wang) [inline include/AsyncTraceDispatcher.h]: Please add new blank line at the end of this file.

File: include/ConsumeMessageContext.h [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+  std::list<MQMessageExt> msgList;
+  MQMessageQueue mq;
+  bool success;
+  std::string status;
+  TraceContext* mqTraceContext;
+  std::map<std::string, std::string> props;
+  std::string msgnamespace;
+
+ public:
+  std::string getConsumerGroup() { return consumerGroup; };
```
Comment (jovany-wang) [inline include/ConsumeMessageContext.h]:
```suggestion
  std::string getConsumerGroup() const { return consumerGroup; };
```

File: include/ConsumeMessageContext.h [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+
+  void setConsumerGroup(std::string consumerGroup) { consumerGroup = consumerGroup; };
+
+  std::list<MQMessageExt> getMsgList() { return msgList; };
```
Comment (jovany-wang) [inline include/ConsumeMessageContext.h]:
```suggestion
  std::list<MQMessageExt> getMsgList() const { return msgList; };
```

File: include/ConsumeMessageContext.h [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+
+  void setMsgList(std::list<MQMessageExt> msgList) { msgList = msgList; };
+  void setMsgList(std::vector<MQMessageExt> pmsgList) { msgList.assign(pmsgList.begin(), pmsgList.end()); };
+  MQMessageQueue getMq() { return mq; };
```
Comment (jovany-wang) [inline include/ConsumeMessageContext.h]:
```suggestion
  MQMessageQueue getMq() const { return mq; };
```

File: include/ConsumeMessageHook.h [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+#define __CONSUMEMESSAGEHOOK_H__
+
+#include <string>
+#include "ConsumeMessageContext.h"
+
+namespace rocketmq {
+
+class ConsumeMessageHook {
+ public:
+  virtual std::string hookName() { return ""; }
```
Comment (jovany-wang) [inline include/ConsumeMessageHook.h]: hookName -> getHookname?

**Comment by ShannonDing** [comment]
the code was refactored in #276, this PR will be closed.
the message trace,  for the async sending, is not supported now, and welcome to add this functionality.

**Pull request closed without merging**